aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp

```diff
@@ -0,0 +1,110 @@
+#include <ATen/ATen.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/native/nestedtensor/NestedTensorImpl.h>
+#include <c10/core/DispatchKey.h>
+
+namespace at {
+namespace native {
+
+NestedTensorImpl::NestedTensorImpl(
+    at::Tensor buffer,
+    at::Tensor nested_size_tensor)
+    : TensorImpl(
+          c10::DispatchKeySet({DispatchKey::NestedTensor}),
+          buffer.dtype(),
+          buffer.device()),
+      _buffer(std::move(buffer)),
+      _nested_size_tensor(std::move(nested_size_tensor)) {
+  remove_autograd_key();
+  key_set_ =
+      key_set_ - c10::DispatchKeySet({c10::DispatchKey::ADInplaceOrView});
+}
+
+at::Tensor wrap_buffer(at::Tensor buffer, at::Tensor nested_size_tensor) {
+  TORCH_CHECK(buffer.is_contiguous(), "Given buffer must be contiguous.");
+  return at::detail::make_tensor<NestedTensorImpl>(
+      std::move(buffer), std::move(nested_size_tensor));
+}
+
+std::vector<at::Tensor> NestedTensor_unbind(
+    const at::Tensor& self,
+    int64_t dim) {
+  TORCH_CHECK(
+      dim == 0,
+      "NestedTensor can only be unbound along dimension 0 ",
+      "got dimension ",
+      dim,
+      " instead.");
+  auto esizes = get_nested_size_tensor(self);
+  std::vector<at::Tensor> result_tensors;
+  if (esizes.dim() == 0) {
+    return result_tensors;
+  }
+  auto esizes_chunks = esizes.unbind(0);
+  std::vector<int64_t> splits;
+  for (const auto i : c10::irange(esizes_chunks.size())) {
+    splits.push_back(esizes_chunks[i].prod().item<int64_t>());
+  }
```

Comment:
This algorithm here feels like there ought to be a kernel that can do it directly, rather than unbinding it and doing it manually yourself. Like, shouldn't this be `torch.prod(esizes, dim=1)` (and maybe a dtype conversion to long) and then you can just view the memory as an IntArrayRef directly without having to allocate the vector.

Reply:
esize_chunks is used again below but the use there also feels unnecessary

Reply:
Let's clean this up as part of profile-guided optimization based on use cases.

Reply:
For now I want to aim for correctness and development speed. There is a lot we can do here, including finding a more efficient data structure to store these numbers to begin with.

Reply:
Yeah, I'm heuristic'ing here "ops involving small tensors are antipattern". It's OK to defer to swolchok here I guess.